### ISSUE TITLE
[Bug] thisCompilation.mainTemplate is never hit

### ISSUE THREAD

**Issue by Craga89** [waysact/webpack-subresource-integrity#14]
Using this plugin in our webpack build produces no difference to the output `webpackJsonp` function. After some debugging, it looks as if the code within the `this-compilation` plugin is never getting called.

If I change the following lines:

```js
  compiler.plugin('compilation', function compilationPlugin(compilation) {
    /*
     * Double plug-in registration in order to push our
     * plugin to the end of the plugin stack.
     */
    compiler.plugin('this-compilation', function thisCompilationPlugin(thisCompilation) {
      thisCompilation.mainTemplate.apply(new WebIntegrityJsonpMainTemplatePlugin());
    });
```

To this:

```js
  compiler.plugin('compilation', function compilationPlugin(compilation) {
    /*
     * Double plug-in registration in order to push our
     * plugin to the end of the plugin stack.
     */
    compilation.mainTemplate.apply(new WebIntegrityJsonpMainTemplatePlugin());
```

It works.

Looking at the webpack source code, it looks like `this-compilation` event/plugin actually [fires _before_ `compilation`, not after](https://github.com/webpack/webpack/blob/9162f9e6eea62137139f95b8aaedee335c870edd/lib/Compiler.js#L397-L398), so this code will never get hit.

**Comment by jscheid** [comment]
Hi, thanks for your bug report. See also #4 where somebody reported a similar issue, unfortunately back then I wasn't able to reproduce.

Can you provide a failing test case or a sample project that shows the problem? Which version of webpack are you using?

**Comment by jscheid** [comment]
Regarding your analysis of `this-compilation`: firstly the code path definitely can get hit (the plugin works fine in our webpack config, and the test cases all pass -- try uncommenting the code in question and they will fail.) Secondly you're assuming that `Compiler.newCompilation` is only executed once per compiler run, which is not (necessarily) the case as you can verify yourself e.g. by adding some console output to that function and then running our test suite.

That is not to say this isn't a genuine bug, it's the second report of the same issue so clearly something is going on. It's just going to be difficult to fix without being able to reproduce it. Any information you can provide about the versions you're using, the build environment, your webpack config etc would be much appreciated. Again, best would be a sample project that I can build to see the problem, or even better a new test case for our test suite.

**Comment by jscheid** [comment]
I've had a look at the code and I think there's an elegant way to make it work reliably (and better in your case) by using the `after-optimize-assets` entry point instead of the `optimize-assets` entry point and do away with the double registration hack. Still, I'd feel much better making this change if I could reproduce your issue first, so for now I'm going to wait for more info from you.

**Comment by Craga89** [comment]
The specific Webpack build I'm using is a production site for the company I work for but I'll do my best to boil this down to specific versions where I can reproduce. Thanks for the detailed reply!

**Comment by jscheid** [comment]
@Craga89 can you see if it works better when using #16 ?

    npm install waysact/webpack-subresource-integrity#better-plugin-registration

**Comment by Craga89** [comment]
I'll give that a shot tomorrow morning @jscheid  thanks!

**Comment by Craga89** [comment]
@jscheid did you publish the branch? Doesn't seem to be present in the GitHub repo

**Comment by jscheid** [comment]
@Craga89 whoops, it got deleted when I merged the PR. It's on master now, can you try with that?

**Comment by Craga89** [comment]
That fixed it, thanks @jscheid 👍

**Comment by jscheid** [comment]
No problem, thanks for testing. I'm planning to roll a 0.7.0 release with this in the next couple of days.

**Comment by jscheid** [comment]
The fix for this was released in v0.7.0.
